### PR TITLE
Fix bias gradient computations

### PR DESCRIPTION
**Pull request by baberabb**
Couple of bug fixes to do with bias:
  - The gradients were contracted over both the batch and sequence dimension (`dim=(0,1)`), rather than just the sequence (`dim=1`).
  - Normalize weights with Adam before concatenating bias to avoid shape mismatch (`[N, O, I+1] / [O, I]` division error). ~~The biases are currently concatenated raw, as I wasn't sure the best way to handle them. More in comment.~~

  update:

  - Added `bias_avg_sq` field to `AdafactorNormalizer` and `AdamNormalizer` to keep track of the bias second moments so we can handle bias normalization separately from weight gradients in `AdafactorNormalizer.normalize_()`:
    - Normalize bias from raw gradient `G` before weight processing
    - Sum bias gradients over sequence dimension
    - Append normalized bias as extra column when include_bias=True

Modified `GradientCollectorCallback` (with help from claude):
   -  Extract bias second moments from both adam and adafactor optimziers
   -  added `scale_by_lr(lr)` method to `AdafactorNormalizer` (also fixes bug where optimizer state tensors were being modified in-place)
   - added `test_optimizer_state_extraction`

~~Also added some unit tests. #75 should probably be merged before this.~~

Someone better at linear algebra than me should probably have a look at this as well.

### CODE REVIEW

**Comment by luciaquirke** [comment]
This is fabulous, thank you!! 🙏 Interested to hear what Nora thinks but I reckon exposing second moments for bias through the normalizer would be great

**Comment by baberabb** [comment]
> Running
> 
> ```
> pip install -e ".[dev]"
> pre-commit install
> ```
> 
> Should add formatting on commit, let me know if that doesn't work for some reason

oh yeah, it was a problem with the ruff linter. it doesn't fix line length errors (leaves that to the formatter). Will add black back.

File: bergson/gradients.py [hunk 1]
```diff
@@ -157,6 +189,17 @@ def to_adafactor(self) -> AdafactorNormalizer:
         return AdafactorNormalizer(
             row=self.avg_sq.mean(dim=1),  # shape [O]
             col=self.avg_sq.mean(dim=0),  # shape [I]
+            bias_avg_sq=self.bias_avg_sq,  # Preserve bias second moments
+        )
+
+    def scale_by_lr(self, lr: float | Tensor) -> None:
```
[the remark this answers is not in the document]
Comment (baberabb) [reply bergson/gradients.py]: was waffling between the two, hence this cursed syntax. New tensors, since these are references initially, and easy to forget to replace before calling.

File: bergson/collector/gradient_collectors.py [hunk 2]
```diff
@@ -143,13 +145,8 @@ def forward_hook(self, module: nn.Module, a: Float[Tensor, "N S I"]) -> None:
...
             a = a * a_factor.type_as(a)  # [N, S, I] * [I] → [N, S, I]
 
-        if module._has_bias:
-            # Append ones to activation for bias term
-            ones = torch.ones(a.size(0), a.size(1), 1, device=a.device, dtype=a.dtype)
-            a = torch.cat([a, ones], dim=-1)
-            i = i + 1
-            setattr(module, LayerAdapter.in_attr(module), i)
-        if p is not None:
+        # Only project if no bias (bias requires full gradient to be materialized)
```
[the remark this answers is not in the document]
Comment (baberabb) [reply bergson/collector/gradient_collectors.py]: yup. added it back in

File: bergson/collector/gradient_collectors.py [hunk 3]
```diff
@@ -171,25 +168,100 @@ def backward_hook(self, module: nn.Module, g: Float[Tensor, "N S O"]):
...
-            P = normalizer.normalize_(full_gradient)
-            if p is not None:
-                g_projection = self.projection(name, p, o, "left", g.device, g.dtype)
-                a_projection = self.projection(name, p, i, "right", g.device, g.dtype).T
-                P = g_projection @ P @ a_projection
-        else:
-            if isinstance(normalizer, AdafactorNormalizer):
+        bias_grad = None
+
+        match normalizer:
```
[the remark this answers is not in the document]
Comment (LouisYRYJ) [reply bergson/collector/gradient_collectors.py]: I agree that it is quite repetitive. What exactly do you want to upstream? the forward and backward methods? I will be adding instances of HookBaseCollector in near future that will *not* be of this form. But could have a generic GradientCollector thing maybe that is then further inherited?
